File: ai-services/security/pii_scrubber.py
"""
PII Scrubber - Python Implementation
HIPAA-compliant PII redaction for therapeutic transcripts
"""

import re
import random
from dataclasses import dataclass
from enum import Enum
from typing import Any
import logging

logger = logging.getLogger(__name__)

class PIICategory(str, Enum):
    NAMES = "names"
    EMAILS = "emails"
    PHONES = "phones"
    ADDRESSES = "addresses"
    SSN = "ssn"
    DATES = "dates"
    FINANCIAL = "financial"


@dataclass
class ScrubberOptions:
    mask_type: str = "placeholder"  # placeholder, redacted, randomized
    enabled_categories: list[PIICategory] = None
    custom_replacements: dict[str, str] = None

    def __post_init__(self):
        if self.enabled_categories is None:
            self.enabled_categories = [
                PIICategory.NAMES,
                PIICategory.EMAILS,
                PIICategory.PHONES,
                PIICategory.ADDRESSES,
                PIICategory.SSN,
                PIICategory.DATES,
                PIICategory.FINANCIAL,
            ]
        if self.custom_replacements is None:
            self.custom_replacements = {}


PII_PATTERNS: dict[PIICategory, list[re.Pattern]] = {
    PIICategory.NAMES: [
        re.compile(r"\b(?:Mr\.|Ms\.|Mrs\.|Dr\.)\s+[A-Z][a-z]+(?:\s+[A-Z][a-z]+)*\b"),
    ],
    PIICategory.EMAILS: [
        re.compile(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b"),
    ],
    PIICategory.PHONES: [
        re.compile(r"\b(?:\+?\d{1,3}[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b"),
    ],
    PIICategory.ADDRESSES: [
        re.compile(r"\b\d{1,5}\s+[A-Z][a-z]+(?:\s+[A-Z][a-z]+)*\s+(?:St|Ave|Rd|Blvd|Ln|Ct|Dr)\b"),
    ],
    PIICategory.SSN: [
        re.compile(r"\b\d{3}-\d{2}-\d{4}\b"),
    ],
    PIICategory.DATES: [
        re.compile(r"\b(?:\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4}|(?:[A-Z][a-z]+\s+\d{1,2},\s+\d{4}))\b"),
    ],
    PIICategory.FINANCIAL: [
        re.compile(r"\b(?:\d{4}-){3}\d{4}\b"),  # Credit card format
    ],
}

PLACEHOLDERS: dict[PIICategory, str] = {
    PIICategory.NAMES: "[NAME]",
    PIICategory.EMAILS: "[EMAIL]",
    PIICategory.PHONES: "[PHONE]",
    PIICategory.ADDRESSES: "[ADDRESS]",
    PIICategory.SSN: "[SSN]",
    PIICategory.DATES: "[DATE]",
    PIICategory.FINANCIAL: "[FINANCIAL]",
}


def scrub_pii(text: str, options: ScrubberOptions | None = None) -> str:
    """Redacts PII from text based on configured categories"""
    # Emit HIPAA audit event for PHI processing only when text is non‑empty
    if text:
        ProtectedHealthData.emit_audit_event("PII_PROCESSING", {"function": "scrub_pii"})
    
    if not text:
        return text

    if options is None:
        options = ScrubberOptions()

    # Emit audit event before processing
    audit_logger = logging.getLogger('hipaa.audit')
    audit_logger.info(
        "PHI_ACCESS|accessor=%s|categories=%s|mask_type=%s|original_count=%d",
        options.get('accessor', 'unknown'),
        [c.value for c in options.enabled_categories],
        options.mask_type,
        len(text)  # rough indicator of original size
    )

    scrubbed_text = text

    # Apply custom replacements first
    for target, replacement in options.custom_replacements.items():
        scrubbed_text = scrubbed_text.replace(target, replacement)

    # Apply category-based scrubbing
    for category in options.enabled_categories:
        patterns = PII_PATTERNS.get(category, [])
        for pattern in patterns:
            if options.mask_type == "placeholder":
                scrubbed_text = pattern.sub(PLACEHOLDERS[category], scrubbed_text)
            elif options.mask_type == "redacted":
                scrubbed_text = pattern.sub("[REDACTED]", scrubbed_text)
            elif options.mask_type == "randomized":
                # Generate a short random string to avoid leaking length information
                random_replacement = ''.join(random.choice('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789') for _ in range(4))
                scrubbed_text = pattern.sub(lambda m: random_replacement, scrubbed_text)
            else:
                raise ValueError(f"Unsupported mask_type: {options.mask_type}")

    # Emit audit event with result summary
    audit_logger.info(
        "PHI_SCRUB_RESULT|categories=%s|mask_type=%s|scrubbed_length=%d",
        [c.value for c in options.enabled_categories],
        options.mask_type,
        len(scrubbed_text)
    )

    return scrubbed_text


def scan_for_pii(text: str) -> dict[str, Any]:
    """Scans text for PII without modifying it"""
    # Emit HIPAA audit event for PHI processing
    ProtectedHealthData.emit_audit_event("PII_PROCESSING", {"function": "scan_for_pii"})
    
    result = {
        "found": False,
        "categories": [],
        "count": 0,
    }

    for category, patterns in PII_PATTERNS.items():
        for pattern in patterns:
            matches = pattern.findall(text)
            if matches:
                result["found"] = True
                result["categories"].append(category.value)
                result["count"] += len(matches)

    # Emit audit event for scan activity
    audit_logger = logging.getLogger('hipaa.audit')
    audit_logger.info(
        "PHI_SCAN|found=%s|categories=%s|match_count=%d",
        result["found"],
        result["categories"],
        result["count"]
    )

    return result


if __name__ == "__main__":
    # Test
    sample = "Contact Dr. John Smith at john@example.com or (555) 010-9988"
    # Use logging instead of print for production code
    logging.basicConfig(level=logging.INFO)
    logger.info("Scrubbed sample: %s", scrub_pii(sample))
    logger.info("Scan result: %s", scan_for_pii(sample))
    # Original data is not logged to avoid exposing PHI

File: ai-services/security/pii-scrubber.ts
/**
 * PII Scrubber
 *
 * Utility for redacting Personally Identifiable Information (PII) from therapeutic transcripts.
 * Essential for HIPAA compliance and ensuring psychological safety in training sessions.
 */

import { createBuildSafeLogger } from '../../logging/build-safe-logger'
import { requireConsent } from '../../security/consent'

const logger = createBuildSafeLogger('PIIScrubber')

export interface ScrubberOptions {
    maskType?: 'placeholder' | 'redacted' | 'randomized'
    enabledCategories?: PIICategory[]
    customReplacements?: Record<string, string>
}

export type PIICategory =
    | 'names'
    | 'emails'
    | 'phones'
    | 'addresses'
    | 'ssn'
    | 'dates'
    | 'financial'

const PII_PATTERNS: Record<PIICategory, RegExp> = {
    names: /\b(?:Mr\.|Ms\.|Mrs\.|Dr\.)\s+[A-Z][a-z]+(?:\s+[A-Z][a-z]+)*\b/g, // Simple name detection
    emails: /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g,
    phones: /\b(?:\+?\d{1,3}[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b/g,
    addresses: /\b\d{1,5}\s+[A-Z][a-z]+(?:\s+[A-Z][a-z]+)*\s+(?:St|Ave|Rd|Blvd|Ln|Ct|Dr)\b/g,
    ssn: /\b\d{3}-\d{2}-\d{4}\b/g,
    dates: /\b(?:(?:\d{1,2}[/-]\d{1,2}[/-]\d{2,4})|(?:[A-Z][a-z]+\s+\d{1,2},\s+\d{4}))\b/g,
    financial: /\b(?:\d{4}-){3}\d{4}\b/g, // Credit card format
}

const PLACEHOLDERS: Record<PIICategory, string> = {
    names: '[NAME]',
    emails: '[EMAIL]',
    phones: '[PHONE]',
    addresses: '[ADDRESS]',
    ssn: '[SSN]',
    dates: '[DATE]',
    financial: '[FINANCIAL]',
}

/**
 * Generates a randomized placeholder for a given PII category.
 * Appends a random numeric suffix to make each redaction unique.
 */
function getRandomizedPlaceholder(category: PIICategory): string {
    const base = PLACEHOLDERS[category];
    const randomSuffix = Math.floor(Math.random() * 1000).toString();
    return `${base.replace(/[\[\]]/g, '')}${randomSuffix}`;
}

/**
 * Redacts PII from text based on configured categories
 */
export function scrubPII(text: string, options: ScrubberOptions = {}): string {
    if (!text) return text

    // Ensure PHI processing is consented and audited before touching the text
    ensurePHIProcessingAllowed('scrubPII')

    const {
        maskType = 'placeholder',
        enabledCategories = ['names', 'emails', 'phones', 'ssn'],
        customReplacements = {},
    } = options

    // Validate maskType
    const validMaskTypes = ['placeholder', 'redacted', 'randomized'] as const
    if (!validMaskTypes.includes(maskType as any)) {
        throw new Error(`Invalid maskType: "${maskType}". Valid values are: ${validMaskTypes.join(', ')}`)
    }

    let scrubbedText = text

    // Apply custom replacements first
    for (const [target, replacement] of Object.entries(customReplacements)) {
        const escapedTarget = target.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
        scrubbedText = scrubbedText.replace(new RegExp(escapedTarget, 'g'), replacement)
    }

    // Apply category-based scrubbing
    for (const category of enabledCategories) {
        const pattern = PII_PATTERNS[category]
        if (!pattern) continue

        if (maskType === 'placeholder') {
            scrubbedText = scrubbedText.replace(pattern, PLACEHOLDERS[category])
        } else if (maskType === 'redacted') {
            scrubbedText = scrubbedText.replace(pattern, '[REDACTED]')
        } else if (maskType === 'randomized') {
            scrubbedText = scrubbedText.replace(pattern, getRandomizedPlaceholder(category))
        } else {
            // Fail fast for unsupported mask types to avoid silent data leakage
            throw new Error(`Unsupported maskType: ${maskType}. Allowed values are 'placeholder', 'redacted', 'randomized'.`);
        }
    }

    // Validate maskType to catch future typos early
    if (!['placeholder', 'redacted', 'randomized'].includes(maskType)) {
        throw new Error(`Unsupported maskType: "${maskType}". Valid values are "placeholder", "redacted", or "randomized".`);
    }

    // Emit HIPAA audit event for PHI access
    logger.info('HIPAA audit: scrubPII accessed text', {
        accessedAt: new Date().toISOString(),
        function: 'scrubPII',
        categories: enabledCategories,
        originalLength: text.length,
        newLength: scrubbedText.length,
    })

    logger.debug('Text scrubbed for PII', {
        originalLength: text.length,
        newLength: scrubbedText.length,
        categoriesUsed: enabledCategories,
    });

    return scrubbedText;
}

/**
 * Scans text for PII without modifying it
 */
export function scanForPII(text: string): {
    found: boolean
    categories: PIICategory[]
    count: number
} {
    // Verify consent before accessing PHI
    requireConsent('scanForPII')

    const result = {
        found: false,
        categories: [] as PIICategory[],
        count: 0,
    };

    for (const [category, pattern] of Object.entries(PII_PATTERNS)) {
        const matches = text.match(pattern);
        if (matches && matches.length > 0) {
            result.found = true;
            result.categories.push(category as PIICategory);
            result.count += matches.length;
        }
    }

    // Emit HIPAA audit event for PHI access
    logger.info('HIPAA audit: scanForPII accessed text', {
        accessedAt: new Date().toISOString(),
        function: 'scanForPII',
        categories: result.categories,
        matchCount: result.count,
    })

    return result
}